src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -86,12 +88,31 @@ public class TopicPartitionChannel {
   // This offset is updated when Snowflake has received offset from insertRows API
   // We will update this value after calling offsetToken API for this channel
   // We will only update it during start of the channel initialization
-  private long offsetPersistedInSnowflake = NO_OFFSET_TOKEN_REGISTERED_IN_SNOWFLAKE;
+  private AtomicLong offsetPersistedInSnowflake =
+      new AtomicLong(NO_OFFSET_TOKEN_REGISTERED_IN_SNOWFLAKE);
 
   // used to communicate to the streaming ingest's insertRows API
   // This is non final because we might decide to get the new instance of Channel
   private SnowflakeStreamingIngestChannel channel;
 
+  /**
+   * Offsets are reset in kafka when one of following cases arises in which we rely on source of
+   * truth (Which is Snowflake's committed offsetToken)
+   *
+   * <p>1. If channel fails to fetch offsetToken in kafka, we reopen the channel and try to fetch
+   * offset from Snowflake again
+   *
+   * <p>2. If channel fails to ingest a buffer(Buffer containing rows/offsets), we reopen the
```

Comment:
When we say "ingest" here do we mean in the Client SDK side of things or in the Snowflake side of things? If the Client SDK rejects a set of rows we should back off rather than immediately re-opening the channel.

Reply:
I don't think it differentiates at the moment. It will just reopen the channel in case of SFException. SFException is overloaded in the sense it is sent as exception when error occurs from both sides. 

Ideally we should backoff only if it sends 429?